Math/NumberTheory/ArithmeticFunctions/SieveBlock/Unboxed.hs

```diff
@@ -0,0 +1,118 @@
+-- |
+-- Module:      Math.NumberTheory.ArithmeticFunctions.SieveBlock.Unboxed
+-- Copyright:   (c) 2017 Andrew Lelechenko
+-- Licence:     MIT
+-- Maintainer:  Andrew Lelechenko <andrew.lelechenko@gmail.com>
+-- Stability:   Provisional
+-- Portability: Non-portable (GHC extensions)
+--
+-- Bulk evaluation of arithmetic functions without factorisation
+-- of arguments.
+--
+
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE MagicHash           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UnboxedTuples       #-}
+
+module Math.NumberTheory.ArithmeticFunctions.SieveBlock.Unboxed
+  ( SieveBlockConfig(..)
+  , sieveBlockUnboxed
+  ) where
+
+import Control.Monad
+import Control.Monad.ST
+import qualified Data.Vector.Unboxed as V
+import qualified Data.Vector.Unboxed.Mutable as MV
+import GHC.Exts
+
+import Math.NumberTheory.Primes
+import Math.NumberTheory.Powers.Squares
+import Math.NumberTheory.Utils (splitOff#)
+import Math.NumberTheory.Utils.FromIntegral
+
+-- | A record of arguments, which specifies both a function to evaluate
+-- and block's bounds.
+--
+-- For example, here is a configuration for the totient function on block [1..1000]:
+--
+-- > SieveBlockConfig
+-- >   { sbcEmpty                = 1
+-- >   , sbcAppend               = (*)
+-- >   , sbcFunctionOnPrimePower = (\p a -> (p - 1) * p ^ (a - 1)
+-- >   , sbcBlockLowBound        = 1
+-- >   , sbcBlockLength          = 1000
+-- >   }
+data SieveBlockConfig a = SieveBlockConfig
+  { sbcEmpty                :: a
+    -- ^ value of a function on 1
+  , sbcAppend               :: a -> a -> a
+    -- ^ how to combine values of a function on coprime arguments
+  , sbcFunctionOnPrimePower :: Word -> Word -> a
+    -- ^ how to evaluate a function on prime powers
```

Comment:
When I first read this it took me a bit to figure out what role each of these functions play. I feel like `sbcEmpty`, `sbcFunctionOnPrimePower`, `sbaAppend` is a more natural order since it puts the base cases together (unless the current order is more natural for currying and I just don't realize it yet)

Reply:
Actually, it feels odd having these functions and the evaluation range be part of the same object, in general